File: 3/3.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
	"strconv"
	"strings"
)

var claimRegex = regexp.MustCompile(`\A#(\d+)\s@\s(\d+),(\d+):\s(\d+)x(\d+)\z`)

type claim struct {
	id   int
	left int
	top  int
	wide int
	tall int
}

func (c claim) String() string {
	return fmt.Sprintf("id: %d, left: %d, top: %d, wide: %d, tall: %d",
		c.id, c.left, c.top, c.wide, c.tall)
}

func newClaim(claimStr string) (*claim, error) {
	groups := claimRegex.FindAllStringSubmatch(claimStr, -1)
	var res []int
	for i, s := range groups[0] {
		if i == 0 {
			continue
		}
		n, err := strconv.Atoi(s)
		if err != nil {
			return nil, err
		}
		res = append(res, n)
	}
	return &claim{
		id:   res[0],
		left: res[1],
		top:  res[2],
		wide: res[3],
		tall: res[4],
	}, nil
}

type fabric struct {
	width           int
	height          int
	canvas          []int
	claims          map[int][]int
	overlappedCells int
}

func newFabric(width, height int) *fabric {
	return &fabric{
		width,
		height,
		make([]int, width*height),
		make(map[int][]int),
		0,
	}
}

func (f *fabric) addClaim(c claim) {
	cells := f.getClaimCells(c)
	for _, cell := range cells {
		f.canvas[cell]++
		if f.canvas[cell] == 2 {
			f.overlappedCells++
		}
	}
	f.claims[c.id] = cells
}

func (f *fabric) getClaimCells(c claim) []int {
	var cells []int
	for i := 0; i < c.tall; i++ {
		curHeight := f.height * (c.top + i)
		rowStart := curHeight + c.left
		rowEnd := rowStart + c.wide
		for cell := rowStart; cell < rowEnd; cell++ {
			cells = append(cells, cell)
		}
	}
	return cells
}

func (f *fabric) disjointClaimID() int {
	for claimID, cells := range f.claims {
		found := true
		for _, cell := range cells {
			if f.canvas[cell] > 1 {
				found = false
				break
			}
		}
		if found {
			return claimID
		}
	}
	return -1
}

func main() {
	b, err := ioutil.ReadFile("input.txt")
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	const fabricSize = 1000
	f := newFabric(fabricSize, fabricSize)
	claims := strings.Split(string(b), "\n")
	for _, claimStr := range claims {
		claim, err := newClaim(claimStr)
		if err != nil {
			fmt.Printf("could not create claim from str: %v", err)
			os.Exit(1)
		}
		f.addClaim(*claim)
	}

	fmt.Printf("overlapping cells: %d\n", f.overlappedCells)
	fmt.Printf("disjoint claim id: %d\n", f.disjointClaimID())
}

File: 3/3.py
import re
from collections import defaultdict

class Claim:
    claim_re = re.compile('\A#(\d+)\s@\s(\d+),(\d+):\s(\d+)x(\d+)\Z')

    def __init__(self, claim_str):
        groups = Claim.claim_re.match(claim_str).groups()
        assert(len(groups) == 5)

        self.id = int(groups[0])
        self.left = int(groups[1])
        self.top = int(groups[2])
        self.wide = int(groups[3])
        self.tall = int(groups[4])

    def __str__(self):
        return 'id: {}, left: {}, top: {}, wide: {}, tall: {}'.format(
            self.id, self.left, self.top, self.wide, self.tall
        )
    
class Fabric:
    def __init__(self, width, height):
        self.width = width
        self.height = height
        self._canvas = [0] * (width * height)
        self._claims = defaultdict(list)
    
    """
    adds all cells occupied by the claim to this fabric's canvas
    """
    def add_claim(self, claim):
        for cell in self._get_claim_cells(claim):
            self._claims[claim.id].append(cell)
            self._canvas[cell] += 1

    """
    returns the number of cells that are shared between two or more claims
    """
    def get_num_overlapped_cells(self):
        return sum([1 for x in self._canvas if x > 1])

    """
    finds a claim that has no overlapping cell with any other claims
    """
    def find_disjoint_claim(self):
        for claim_id, cells in self._claims.items():
            for cell in cells:
                if self._canvas[cell] != 1:
                    break
            else:
                return claim_id
        return -1

    def clear(self):
        self._canvas = [0] * (self.width * self.height)
        self._claims = defaultdict(list)

    """
    computes all cells the claim occupies
    """
    def _get_claim_cells(self, claim):
        for i in range(claim.tall):
            curr_height = self.height * (claim.top + i)
            row_start = curr_height + claim.left
            row_end = row_start + claim.wide
            for cell in range(row_start, row_end):
                assert(cell <= self.height * self.width - 1)
                yield cell

with open('input.txt') as f:
    claims_input = f.read().splitlines()

FABRIC_SIZE = 1000
fabric = Fabric(FABRIC_SIZE, FABRIC_SIZE)
for claim_str in claims_input:
    claim = Claim(claim_str)
    fabric.add_claim(claim)

print('overlapping cells:', fabric.get_num_overlapped_cells())
print('disjoint claim id:', fabric.find_disjoint_claim())